### ISSUE TITLE
RIndex thoughts

### ISSUE THREAD

**Issue by echatav** [VinylRecords/Vinyl#88]
Hi @jonsterling!

So I was thinking about the `RIndex` type and it seems to be a bit problematic. First it only allows the first appearance of a value to be indexed. Second, if a value isn't in the record it will just keep succeeding. I could be totally wrong in my understanding. But, another way to index would be to smash `Nat` and `RIndex` together.

```haskell
data RIndex rs r where
  Z :: RIndex (r ': rs) r
  S :: RIndex rs r -> RIndex (s ': rs) r
```

These indices are guaranteed to be in the list and if the same `r` appears more than once you can construct all the indices that point to any of them. This would subsume `RElem`.

```haskell
rget :: RIndex rs r -> Rec f rs ->  f r
rget Z (r :& rs) = r
rget (S n) (r :& rs) = rget n rs
```

`RImage` becomes a record itself.

```haskell
type RImage rs ss = Rec (RIndex rs) ss
```

It even forms a category! Although, I can't figure out how to convince GHC it's a `Category` >:-(

```haskell
class IdRImage rs where
  idRImage :: RImage rs rs
instance IdRImage '[] where
  idRImage = RNil
instance IdRImage rs => IdRImage (r ': rs) where
  idRImage = Z :& rmap S idRImage
(.&) :: RImage ss ts -> RImage rs ss -> RImage rs ts
(.&) Nil = pure Nil
(.&) (n :& ns) = (:&) <$> rget n <*> (.&) ns
```

This is not better but different. Instead of asking "here's an `r`, what's its index?", it asks "here's an index, what's its `r`?" I think when GHC 8 unifies types and kinds, you can recover the former semantics, just a bit more safely with the same definition?

```haskell
type family IndexOf r rs where
  IndexOf r (r ': rs) = Z
  IndexOf r (s ': rs) = S (IndexOf r rs)
```

Miss working with ya bud, hope grad school is awesome!

**Comment by acowley** [comment]
Hey, @echatav, sorry I didn't respond earlier! You're right about `RIndex`'s partiality; it ends up working out because client code demands a type class instance, and that search fails when appropriate.

I like how your take on it simplifies `RImage`; have you pushed the change through to see if anything else would be affected?

**Comment by echatav** [comment]
No, I didn't yet, but I'm using these types in work code. The idea comes from the `Elem` type from [glambda](https://hackage.haskell.org/package/glambda-1.0.1/docs/Language-Glambda-Exp.html#t:Elem). I'll give it a shot and see what it looks like in vinyl.

**Comment by echatav** [comment]
Hi @acowley, this should replace most of the current indexing and lens functionality with slightly simpler and more general functionality. Since `RImage` is replaced with something more general/less constrained (`Rec (RIndex _) _` allows reordering and repetition of elements) I changed the names to be simple plurals. The less general/more constrained functionality can be re-obtained by adding an `RSubset` constraint as desired. If you want to see how this would look in an actual PR I can do that but can you look over it first and see if it doesn't look like something is missing or wrong?

```haskell
{-# LANGUAGE DataKinds             #-}
{-# LANGUAGE GADTs                 #-}
{-# LANGUAGE PolyKinds             #-}
{-# LANGUAGE TypeFamilies          #-}
{-# LANGUAGE TypeOperators         #-}

module Data.Vinyl.Index where

import GHC.Prim
import Data.Vinyl.Core

data RIndex (rs :: [k]) (r :: k) where
  Z :: RIndex (r ': rs) r
  S :: RIndex rs r -> RIndex (s ': rs) r

rget :: RIndex rs r -> Rec f rs -> f r
rget Z (r:&_) = r
rget (S ix) (_:&rs) = rget ix rs

(!) :: Rec f rs -> RIndex rs r -> f r
infixl 9 !
(!) = flip rget

rput :: RIndex rs r -> f r -> Rec f rs -> Rec f rs
rput Z r' (_:&rs) = r':&rs
rput (S ix) r' (r:&rs) = r :& rput ix r' rs

rlens :: Functor g => RIndex rs r -> (f r -> g (f r)) -> Rec f rs -> g (Rec f rs)
rlens Z f (r:&rs) = fmap (:& rs) (f r)
rlens (S ix) f (r:&rs) = fmap (r :&) (rlens ix f rs)

class Rid rs where rid :: Rec (RIndex rs) rs
instance Rid '[] where rid = RNil
instance Rid rs => Rid (r ': rs) where rid = Z :& rmap S rid

-- | specializes as composition of records of indices
rgets :: Rec (RIndex rs) ss -> Rec f rs -> Rec f ss
rgets ixs rec = rmap (rec !) ixs

rputs :: Rec (RIndex rs) ss -> Rec f rs -> Rec f ss -> Rec f rs
rputs RNil rs _ = rs
rputs (ix:&ixs) rs (s:&ss) = rput ix s (rputs ixs rs ss)

rlenses :: Functor g => Rec (RIndex rs) ss -> (Rec f ss -> g (Rec f ss)) -> Rec f rs -> g (Rec f rs)
rlenses ixs = lens (rgets ixs) (rputs ixs)
  where
    lens sa sbt afb s = fmap (sbt s) $ afb (sa s)

type family RElem (rs :: [k]) (r :: k) :: Constraint where
  RElem (r ': rs) r = ()
  RElem (r ': rs) r' = RElem rs r'

type family RSubset (rs :: [k]) (ss :: [k]) :: Constraint where
  RSubset '[] ss = ()
  RSubset (r ': rs) (r ': ss) = RSubset rs ss
  RSubset rs (r ': ss) = RSubset rs ss
```

**Comment by echatav** [comment]
Version of `RElem` and `RSubset` constraints as type classes with List-like `elemIndex` and `elemIndices` functionality:

```haskell
class RElem (r :: k) (rs :: [k]) where relemIndex :: RIndex rs r
instance r `RElem` (r ': rs) where relemIndex = Z
instance (r `RElem` rs) => r `RElem` (r' ': rs) where relemIndex = S relemIndex

rlensFirst
  :: (RElem r rs , Functor g)
  => (f r     -> g (f r))
  -> Rec f rs -> g (Rec f rs)
rlensFirst = rlens relemIndex

rgetFirst :: RElem r rs => Rec f rs -> f r
rgetFirst = rget relemIndex

rputFirst :: RElem r rs => Rec f rs -> f r -> Rec f rs
rputFirst = rput relemIndex

class RSubset (ss :: [k]) (rs :: [k]) where relemIndices :: Rec (RIndex rs) ss
instance '[] `RSubset` rs where relemIndices = RNil
instance (ss `RSubset` rs) => (r ': ss) `RSubset` (r ': rs) where relemIndices = Z :& relemIndices
instance {-# OVERLAPPABLE #-} (rs `RSubset` ss) => rs `RSubset` (r ': ss) where relemIndices = rmap S relemIndices

rsubset
  :: (Functor g, RSubset ss rs)
  => (Rec f ss -> g (Rec f ss))
  ->  Rec f rs -> g (Rec f rs)
rsubset = rlenses relemIndices

rcast :: RSubset ss rs => Rec f rs -> Rec f ss
rcast = rgets relemIndices

rreplace :: RSubset ss rs => Rec f rs -> Rec f ss -> Rec f rs
rreplace = puts relemIndices
```

**Comment by echatav** [comment]
You can also generalize these Peano indices so that in addition to being constructive proofs that `r` is in `rs`, it proves that if you replace `r` by `s` you get `ss`.

```haskell
data RIndex (rs :: [k]) (ss :: [k]) (r :: k) (s :: k) where
  Z :: RIndex (r ': rs) (s ': rs) (r :: k) (s :: k)
  S :: RIndex rs ss r s -> RIndex (t ': rs) (t ': ss) r s
```

Then you can get all the benefits of type-changing lens families which seems pretty useful for `Rec`.

**Comment by acowley** [comment]
I'm failing to come up with a hypothetical argument against doing this, but I'd like to push the change through vinyl and frames at least to see if any difference pops up before switching over. I'm not going to be able to get to that until Thursday, though; please don't take that delay as a lack of interest! Thanks for continuing to push on the idea.

**Comment by echatav** [comment]
You might be interested in supporting relational algebra in frames. Simple sketch:

```haskell
data Relation t f as where
  Table :: t (Rec f as) -> Relation t f as
  Union :: Relation t f as -> Relation t f as -> Relation t f as
  Join :: Relation t f as -> Relation t f bs -> Relation t f (as ++ bs)
  Project :: Rec (RIndex as) bs -> Relation t f as -> Relation t f bs
```

Another thought I was having is that there ought to be a simultaneous generalization of `Rec f bs` and `CoRec g as`, `ProRec p as bs`, in analogy to how profunctors generalize functors and contravariant functors. I'd need something like `ProRec` to construct type-changing records of indices, `ProRec (RIndexFamily ss ts) as bs` and a corresponding lens. Sorry for the mind-dumping!

**Comment by acowley** [comment]
Can you put this into a buildable PR? I spent some time trying to shoehorn it in, but there is too much breakage to quickly evaluate. If you have it mostly working already, the thing I'd look at for an early test as to whether or not this can fly is the existing benchmark in Vinyl. The recursive definitions for `rlens`, `rget`, etc. could be a performance problem.

**Comment by echatav** [comment]
Made a PR. I couldn't get benchmarks to compile. Something about LLVM. I don't think recursive definitions should cause performance issues as indexing `Rec` is necessarily O(n) anyway. I renamed functions to make this as small a change in API as possible, mostly just `Vinyl.Lens` ~> `Vinyl.Index`. I neglected to write documentation, inline directives or rewrite rules. The latter two might impact benchmarks.

**Comment by sboosali** [comment]
@echatav you can temporarily disable `-fllvm` in the benchmarks. Performance is good to know for both backends anyways.

**Comment by acowley** [comment]
@sboosali If you have some time to help with the [PR](https://github.com/VinylRecords/Vinyl/issues/93), I'd be very thankful! I think both @echatav and I ran out of time to put into it back when this discussion happened, but I like his refactoring of the design.

**Comment by echatav** [comment]
Hey, thanks for renewing interest in this. I did run out of time back when we discussed these changes but my time is more flexible now and I'd be down to help out to!

**Comment by acowley** [comment]
I've been thinking about type changing things recently and remembered this `ProRec` construction, but I'm not clear on how it would be used. Did you ever have a chance to push through a type-changing field update function, @echatav?

The comparison of corecords/records to `Star`/`Costar` (or `UpStar`/`DownStar`) is tantalizing.

**Comment by sboosali** [comment]
> The comparison of corecords/records to Star/Costar (or UpStar/DownStar) is tantalizing.

Can you elaborate?

**Comment by echatav** [comment]
@acowley I never got a working notion of ProRec unfortunately, despite trying a few different iterations. But the type-changing index is pretty simple.

```haskell
data Index ss ts a b where
  Z :: Index (s ': ss) (t ': ss) s t
  S :: Index ss ts a b -> Index (s ': ss) (s ': ts) a b

ixLens :: Index ss ts a b -> Lens (Rec f ss) (Rec f ts) (f a) (f b)
ixLens Z f (s:&ss) = fmap (:& ss) (f s)
ixLens (S ix) f (s:&ss) = fmap (s :&) (ixLens ix f ss)
```

So now we want to put together some kind of lens into a whole subset of a record. At first I thought, well, that's just a record of `Simple Index` (although we have a difficulty here with type synonyms' like `type Simple f s a = f s s a a` inability to do partial application). The record would record a list of type changing indices, and their composition of their lenses would give an overall lens. And that does work fine for instance for the `RSubset` machinery. But that clearly only works if the new and old record are the same size, what if I want to drop or add indices? So really, you need two more index types. This ends up looking a lot like the semantics of extensible records in [this paper](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/scopedlabels.pdf). Unfortunately, my initial hope that it would somehow mirror the sort of `Functor`/`Covariant`/`Profunctor` trichotomy didn't seem to pan out :-/

**Comment by echatav** [comment]
If there is an analogy to be had I'm guessing it would be something like:

Rec ~ Product datatype ~ Functor
CoRec ~ Sum datatype ~ Contravariant
ProRec ~ Algebraic datatype ~ Profunctor

**Comment by acowley** [comment]
Yeah, it's similar to products vs sums. We can get a `Profunctor (f a b)` from a `Functor f` with the function `a -> f b` (`Star`), or the function `f a -> b` (`Costar`). If you follow the superficial coincidence, this looks a bit like injection of a field, `(r ∈ rs) => r -> f rs`, and projection of a field, `(r ∈ rs) => f rs -> r`. That is, the relationship of `CoRec` and `Rec` to a particular field.

So then one can't help but think about a field seen as a lens capable of both setting and getting. Say it was a tensor of those two capabilities (injection and projection), one can then pre- or post-compose a function to "change the type" of the field. What doesn't fall out so easily is something we actually care about quite a bit: the aggregate type of all fields in a record (for `vinyl`, a type-level list).

I think `vinyl` has been a useful testbed for this struggle. It started out more focused on index values, then shifted to the typeclass synthesized lenses. @echatav's type-changing index seems to offer something very nice as it bundles up the type-changing capability one could attach to an index with the index itself.

**Comment by echatav** [comment]
The analog for `CoRec` should be a `Prism`

```haskell
ixPrism :: Index ss ts a b -> Prism (CoRec f ss) (CoRec f ts) (f a) (f b)
ixPrism = _
```

**Comment by acowley** [comment]
Yeah, and I'm not sure what else there is to gain by further conceptual unification, but it was a really nice thought prompt!

So we're still at the place where we need to figure out how to make the type changing part go fast.

**Comment by echatav** [comment]
It's not the type-changing aspect which should affect speed. As a roadmap, I'd say defer type-changing to a later release and focus on making the type-constant `RIndex rs r` work fast since this minimizes the API changes (it should have no affect on user code). When that is satisfactory it'd make sense to think of a proper new API or not. Just my 2c. Also, you may want to look at the [type-combinators](https://hackage.haskell.org/package/type-combinators) package where some pieces of the larger picture are worked out.

**Comment by acowley** [comment]
Well, yes I'm familiar with the performance issue. A straight forward presentation of type changing accessors is a good motivator.

Are you also suggesting to use the `type-combinators` package in place of the `TypeLevel` module, or is there something more?

**Comment by acowley** [comment]
I realized I didn't link this before, but [here](https://github.com/VinylRecords/Vinyl/commit/66fadafbff99e680adfb81bcce5ba186ef982479) is an earlier effort at speeding up witness-based record traversals back when that's how things worked. It was fairly effective, but of course arbitrarily limited by how much manual unrolling was done. The subsequent change to typeclass recursion fixed that at a cost of compile time and more uncertainty over exactly how performance scales to larger record sizes.

**Comment by echatav** [comment]
No, just saying it's worth looking at to plumb for ideas, for instance the `Remove` index type helped me. The `Product` type is the same as `Rec` but the difference between `Sum` and `CoRec` seem instructive.

**Comment by acowley** [comment]
Gotcha. I have a `Remove`-like in other packages, but find working with it to be quite clumsy. One ends up with types like `Insert a (Remove b (Remove a rs))` or something. Have you encountered that sort of situation? Kenny Foner has been working a bit on simplifying such types through a compiler plugin, but I'm curious if anyone's found alternate approaches.

The `Sum` difference from `CoRec` comes up [periodically](https://github.com/acowley/Frames/issues/42).

**Comment by echatav** [comment]
Yes, I'd expect it to look awful without sugar :-)

**Comment by acowley** [comment]
Thank you for linking `type-combinators`, btw. Sometimes it is discouraging how much duplication there is on hackage, but that package looks very nicely put together.

**Comment by echatav** [comment]
One thing I was saying on a reddit thread earlier is we should make a mailing list (or whatever) for all the folks working on different "records" solutions; `Vinyl`, `Bookkeeper`, `rawr`, `Dbrecord-opaleye`, `record`, `type-combinators` and `OverloadedRecordFields` and put together everyone's requirements for records and come up with a good solution.

**Comment by echatav** [comment]
Just figured I'd record here the index lens and prism in the profunctor representation:

```haskell
{-# LANGUAGE
    DataKinds
  , GADTs
  , RankNTypes
  , TypeOperators
#-}

module Index where

import Data.Profunctor

type Lens s t a b = forall p. Strong p => p a b -> p s t
type Prism s t a b = forall p. Choice p => p a b -> p s t

data Index ss ts a b where
  Z :: Index (a ': xs) (b ': xs) a b
  S :: Index ss ts a b -> Index (x ': ss) (x ': ts) a b

data Rec f xs where
  Nil :: Rec f '[]
  (:&) :: f x -> Rec f xs -> Rec f (x ': xs)

fromRec :: Rec f (x ': xs) -> (f x, Rec f xs)
fromRec (x:&xs) = (x,xs)

toRec :: (f x, Rec f xs) -> Rec f (x ': xs)
toRec (x,xs) = x:&xs

data CoRec f xs where
  Here :: f x -> CoRec f (x ': xs)
  There :: CoRec f xs -> CoRec f (x ': xs)

fromCoRec :: CoRec f (x ': xs) -> Either (f x) (CoRec f xs)
fromCoRec (Here x) = Left x
fromCoRec (There xs) = Right xs

toCoRec :: Either (f x) (CoRec f xs) -> CoRec f (x ': xs)
toCoRec (Left x) = Here x
toCoRec (Right xs) = There xs

ixLens :: Index ss ts a b -> Lens (Rec f ss) (Rec f ts) (f a) (f b)
ixLens Z = dimap fromRec toRec . first'
ixLens (S ix) = dimap fromRec toRec . second' . ixLens ix

ixPrism :: Index ss ts a b -> Prism (CoRec f ss) (CoRec f ts) (f a) (f b)
ixPrism Z = dimap fromCoRec toCoRec . left'
ixPrism (S ix) = dimap fromCoRec toCoRec . right' . ixPrism ix
```
